roachprod-microbench: cache baseline results

herkolategan opened the pull request:
Previously, we would run microbenchmarks for both sets of binaries (the baseline
and experiment revision). Although this was beneficial to eliminate variance to
some degree, we could rather reuse cached results from the baseline revision and
possibly spend the extra compute on an extra run mid-week.

This change writes the results of the baseline revision to GCS. If the revision
exists in storage it will pull and compare against those results, and only run
the `experiment` revision.

Resolves: #128883

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128295)
<!-- Reviewable:end -->

herkolategan committed:
roachprod-microbench: cache baseline results

Previously, we would run microbenchmarks for both sets of binaries (the baseline
and experiment revision). Although this was beneficial to eliminate variance to
some degree, we could rather reuse cached results from the baseline revision and
possibly spend the extra compute on an extra run mid-week.

This change writes the results of the baseline revision to GCS. If the revision
exists in storage it will pull and compare against those results, and only run
the `experiment` revision.

Epic: None
Release note: None

herkolategan committed:
roachprod-microbench: cache experiment results

The results of the experiment revision is also backed up to GCS, in case we want
to reprocess the data in the future.

Epic: None
Release note: None

herkolategan committed:
roachprod-microbench: save metadata with results

With the results of a run being cached, metadata should be included that give
certain details about the run.

This change stores metadata about the baseline commit, machine type,
architecture and operating system alongside the run results. The repository is
also stored for future reference. The experiment commit also stores its SHA and
commit time in addition to the baseline run metadata.

Epic: None
Release note: None

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -61,17 +73,21 @@ for rev in "${revisions[@]}"; do
 
+# Check if the baseline cache exists and copy it to the output directory.
+baseline_cache_path="gs://$BENCH_BUCKET/cache/$GCE_MACHINE_TYPE/$SANITIZED_BENCH_PACKAGE/${sha_arr[1]}"
+if gsutil ls "$baseline_cache_path" &>/dev/null; then
```
I was going to suggest using `check_gcs_path_exists`, but it seems that the support file is not `source`d here. I actually can't seem to find where that file is used. 🤔

herkolategan replied:
I can probably move this to support, so that it can be reused, although it's such a small function, but probably makes it more clear what it's doing.

renatolabs replied:
Nah, I get it now, I was grepping for the wrong string.

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -61,17 +73,21 @@ for rev in "${revisions[@]}"; do
 
-# Set up credentials (needs to be done after the build phase)
-google_credentials="$GOOGLE_EPHEMERAL_CREDENTIALS"
+# Log into gcloud again (previous credentials are no longer available at this point)
```
What do you mean by no longer available?

herkolategan replied:
This is a silly workaround, `teamcity-support.sh` has a trap that wipes the credentials file on exit, when the build support script above is executed (it also pulls in teamcity-support.sh to login the gcloud in the container, but wipes the outer file).

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -104,10 +120,15 @@ done
-if [ -d "$output_dir/0" ] && [ "$(ls -A "$output_dir/0")" ] \
-&& [ -d "$output_dir/1" ] && [ "$(ls -A "$output_dir/1")" ]; then
+# Compare the results, if both sets of benchmarks were run.
+# These should exists if the benchmarks were run successfully.
```
Nit: `exist`.

herkolategan replied:
Done.

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -75,6 +76,7 @@ done
 
 # Check if the baseline cache exists and copy it to the output directory.
 baseline_cache_path="gs://$BENCH_BUCKET/cache/$GCE_MACHINE_TYPE/$SANITIZED_BENCH_PACKAGE/${sha_arr[1]}"
+baseline_commit=${sha_arr[1]}
```
It took me a while to understand the `unset` call below -- it's a way to signal that we used cached baseline results. But then we erase the baseline SHA, which we now have to copy over here because we actually need it.

Maybe we can signal that we used cached results using some kind of boolean variable instead of array deletion? Bonus for then keeping the symmetry between `sha_arr` and `name_arr`.

herkolategan replied:
True, the logic here is a bit befuddled. To remedy this I opted for a separate array that needs to be built and run, vs. an array to reference the SHAs we're comparing. I tried a boolean, but it felt cleaner using another array.

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -120,6 +122,26 @@ done
+metadata["experiment-commit-time"]=$(git show -s --format=%cI "${sha_arr[0]}")
+for key in "${!metadata[@]}"; do
+  echo "$key": "${metadata[$key]}" >> "$output_dir/experiment/metadata.log"
+done
```
Is the idea that this metadata would be parsed by `roachprod-microbench` in the future? Might be worth having that tool generate the metadata in a well defined format (e.g., JSON) via a data structure in code, rather than writing it in the bash script.

herkolategan replied:
Yes, it's already part of this PR https://github.com/cockroachdb/cockroach/pull/129074. The format is that of the traditional go benchmark logs, and it's what [benchfmt](https://pkg.go.dev/golang.org/x/perf/storage/benchfmt) will consume to add `residues` (as they call it) to the benchmark results. Hence I didn't feel the need to have a tool generate it, but I can add a comment to mention here that these metadata files just become part of the benchmark logs that the reader will consume.

renatolabs reviewed (approved):
A few suggestions, but LGTM!
